File: tools/misc/prepareDev/initBin.rb
require 'fileutils'
require_relative './decompressBin'

BIN_DIR = '_dist/bin'
BIN_PATH_EXE = '_dist/bin/noraneko'
BIN_ROOT_DIR = '_dist/bin'
BIN_VERSION = '_dist/bin/version.txt'
VERSION = '1.0.0'

def init_bin
  has_version = File.exist?(BIN_VERSION)
  has_bin = File.exist?(BIN_PATH_EXE)
  need_init = false

  if has_bin && has_version
    version = File.read(BIN_VERSION).strip
    if VERSION != version
      puts "[dev] Version mismatch: #{version} !== #{VERSION}. Re-extracting."
      FileUtils.rm_rf(BIN_ROOT_DIR)
      need_init = true
    else
      puts '[dev] Binary version matches. No initialization needed.'
    end
  elsif has_bin && !has_version
    puts "[dev] Binary exists but version file is missing. Writing #{VERSION}."
    FileUtils.mkdir_p(BIN_DIR)
    File.write(BIN_VERSION, VERSION)
    puts '[dev] Initialization complete.'
  elsif !has_bin && has_version
    warn '[dev] Version file exists but binary is missing. Abnormal termination.'
    raise 'Unreachable: !has_bin && has_version'
  else
    puts '[dev] Binary not found. Extracting.'
    need_init = true
  end

  if need_init
    FileUtils.mkdir_p(BIN_DIR)
    decompress_bin
    puts '[dev] Initialization complete.'
  end
end

init_bin if $PROGRAM_NAME == __FILE__

File: tools/misc/prepareDev/initBin.ts
import {
  binDir,
  binPathExe,
  binRootDir,
  binVersion,
  VERSION,
} from "../../utils/defines.ts";
import { ensureDir, isExists, safeRemove } from "../../utils/utils.ts";
import { decompressBin } from "./decompressBin.ts";

/**
 * Binary initialization process. Checks version and existence, extracts if needed.
 */
export async function initBin() {
  const hasVersion = await isExists(binVersion);
  const hasBin = await isExists(binPathExe);
  let needInit = false;

  if (hasBin && hasVersion) {
    const version = (await Deno.readTextFile(binVersion)).trim();
    if (VERSION !== version) {
      console.log(
        `[dev] Version mismatch: ${version} !== ${VERSION}. Re-extracting.`,
      );
      await safeRemove(binRootDir);
      needInit = true;
    } else {
      console.log("[dev] Binary version matches. No initialization needed.");
    }
  } else if (hasBin && !hasVersion) {
    console.log(
      `[dev] Binary exists but version file is missing. Writing ${VERSION}.`,
    );
    await ensureDir(binDir);
    await Deno.writeTextFile(binVersion, VERSION);
    console.log("[dev] Initialization complete.");
  } else if (!hasBin && hasVersion) {
    console.error(
      "[dev] Version file exists but binary is missing. Abnormal termination.",
    );
    throw new Error("Unreachable: !hasBin && hasVersion");
  } else {
    console.log("[dev] Binary not found. Extracting.");
    needInit = true;
  }

  if (needInit) {
    await ensureDir(binDir);
    await decompressBin();
    console.log("[dev] Initialization complete.");
  }
}

// Entry point when run directly
if (import.meta.main) {
  initBin().catch((error) => {
    console.error("[dev] Binary initialization failed:", error);
    Deno.exit(1);
  });
}
